aodncore/pipeline/handlerbase.py

```diff
@@ -138,39 +138,123 @@
 
-For example, a handler instantiated with any of these class parameters may also access them from the class instance as
-follows::
+A handler instantiated with any of these class parameters may also access them from the class instance::
 
     from aodncore.pipeline import HandlerBase
     from aodncore.pipeline.config import CONFIG
 
 
     class MyHandler(HandlerBase):
         def print_upload_path(self):
-            # Note: when accessing attributes from within the class itself, the usual Python 'self.attr'
-            # convention applies to access the *current* instance
+            # Note: when accessing attributes from within the class itself, the
+            # usual Python 'self.attr' convention applies to access the *current* instance
             print(self.upload_path)
 
 
-    h = MyHandler('/path/to/input/file.nc', config=CONFIG, upload_path='/original/incoming/path/file.nc')
+    h = MyHandler('/path/to/input/file.nc', config=CONFIG,
+                  upload_path='/original/incoming/path/file.nc')
+
+    # 'input_file' parameter is now available as the 'input_file' attribute
     h.input_file
     '/path/to/input/file.nc'
+
+    # 'upload_path' parameter is now available as the 'upload_path' attribute
     h.upload_path
     '/original/incoming/path/file.nc'
+
+    # 'config' parameter is now available as the 'config' attribute
     h.config
     <aodncore.pipeline.configlib.LazyConfigManager object at 0x7f22230c5990>
 
     h.print_upload_path()
     /original/incoming/path/file.nc
 
+Examples
+--------
+
+Writing a :meth:`dest_path` function
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* Writing a :meth:`dest_path` function with an unmodified filename::
+
+    import os
+
+    class MyHandler(HandlerBase):
+        def dest_path(self, file_path):
+            basename = os.path.basename(file_path)
+            dest_filename = "IMOS_filename_01_XX_{basename}".format(basename=basename)
+            return os.path.join('IMOS/MYFACILITY', dest_filename)
+
+* Writing a :meth:`dest_path` function based on contents of a NetCDF file::
+
+    import os
+    from netCDF4 import Dataset
+
+    class MyHandler(HandlerBase):
+        def dest_path(self, file_path):
+            with Dataset(file_path, mode='r') as d:
+                site_code = d.site_code
+
+            dest_filename = "IMOS_filename_00_{site_code}.nc".format(site_code=site_code)
+            return os.path.join('IMOS/MYFACILITY', dest_filename)
+
+Creating products during the handler lifetime
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* Create a simple product during the :meth:`preprocess` step and add to the file collection::
+
+    import os
+    from aodncore.pipeline import PipelineFile
+
+    class MyHandler(HandlerBase):
+        def preprocess(self):
+            # create the product
+            product_path = os.path.join(self.products_dir, 'product.txt')
+            with open(product_path, 'w') as f:
+                f.write('some file contents' + os.linesep)
```

Comment:
Maybe add a comment to say that usually the product would be created from one or more of the files in the collection, which can be accessed via the `self.collection` attribute

Reply:
I was trying to keep each example to one concept, with the overall assumption being that a real pipeline would combine concepts. I had this rolled together with the NetCDF accessing example, but stripped it back to just show how to add a file.

Could make a note in a comment that the product would typically be created from one of the existing files?